In [1]:
# Cell 1: Import necessary libraries
import pyreadstat
import pandas as pd

In [2]:
#check the current working directory to ensure correct file paths
import os
os.getcwd()

'/Users/ushagartoula/NHANES_DATA'

In [3]:
import os
# Change to your data folder
os.chdir("/Users/ushagartoula/NHANES_DATA")

# Confirm
print(os.getcwd())

/Users/ushagartoula/NHANES_DATA


In [4]:
# Cell 2: Load dietary fiber data
diet, meta_diet = pyreadstat.read_xport("data/DR1TOT_L.xpt", encoding='latin1')
diet_subset = diet[['SEQN', 'DR1TFIBE']]  # SEQN = participant ID, DR1TFIBE = total fiber intake

# Inspect
print(diet_subset.head())
print(diet_subset.isnull().sum())

       SEQN  DR1TFIBE
0  130378.0      10.1
1  130379.0      29.6
2  130380.0      21.1
3  130381.0      11.5
4  130382.0      13.8
SEQN           0
DR1TFIBE    2166
dtype: int64


In [5]:
# Cell 3: Load fasting glucose data
glucose, meta_glu = pyreadstat.read_xport("data/GLU_L.xpt", encoding='latin1')
glucose_subset = glucose[['SEQN', 'LBXGLU']]  # SEQN = participant ID, LBXGLU = glucose

print(glucose_subset.head())
print(glucose_subset.isnull().sum())

       SEQN  LBXGLU
0  130378.0   113.0
1  130379.0    99.0
2  130380.0   156.0
3  130386.0   100.0
4  130394.0    88.0
SEQN        0
LBXGLU    324
dtype: int64


In [6]:
# Cell 4: Load demographics
demo, meta_demo = pyreadstat.read_xport("data/DEMO_L.xpt", encoding='latin1')
demo_subset = demo[['SEQN', 'RIAGENDR', 'RIDAGEYR']]  # Gender and age

print(demo_subset.head())
print(demo_subset.isnull().sum())


       SEQN  RIAGENDR  RIDAGEYR
0  130378.0       1.0      43.0
1  130379.0       1.0      66.0
2  130380.0       2.0      44.0
3  130381.0       2.0       5.0
4  130382.0       1.0       2.0
SEQN        0
RIAGENDR    0
RIDAGEYR    0
dtype: int64


In [7]:
# Cell 5: Load HDL cholesterol
hdl, meta_hdl = pyreadstat.read_xport("data/HDL_L.xpt", encoding='latin1')
hdl_subset = hdl[['SEQN', 'LBDHDD']]  # HDL in mg/dL

print(hdl_subset.head())
print(hdl_subset.isnull().sum())

       SEQN  LBDHDD
0  130378.0    45.0
1  130379.0    60.0
2  130380.0    49.0
3  130386.0    46.0
4  130387.0    42.0
SEQN         0
LBDHDD    1178
dtype: int64


In [8]:
# Cell 6: Load Total Cholesterol data
chol, meta_chol = pyreadstat.read_xport("data/TCHOL_L.xpt", encoding='latin1')

# Subset columns: SEQN = participant ID, LBXTC = total cholesterol
chol_subset = chol[['SEQN', 'LBXTC']]

# Inspect first 5 rows and missing values
print(chol_subset.head())
print(chol_subset.isnull().sum())

       SEQN  LBXTC
0  130378.0  264.0
1  130379.0  214.0
2  130380.0  187.0
3  130386.0  183.0
4  130387.0  203.0
SEQN        0
LBXTC    1178
dtype: int64


In [9]:
# Check all columns in the Triglycerides dataset
trig, meta_trig = pyreadstat.read_xport("data/TRIGLY_L.xpt", encoding='latin1')
print(trig.columns)

Index(['SEQN', 'WTSAF2YR', 'LBXTLG', 'LBDTRSI', 'LBDLDL', 'LBDLDLSI',
       'LBDLDLM', 'LBDLDMSI', 'LBDLDLN', 'LBDLDNSI'],
      dtype='object')


In [10]:
# Load Triglycerides subset using the correct column name
trig_subset = trig[['SEQN', 'LBXTLG']]

# Inspect
print(trig_subset.head())
print(trig_subset.isnull().sum())

       SEQN  LBXTLG
0  130378.0   153.0
1  130379.0    86.0
2  130380.0   375.0
3  130386.0   142.0
4  130394.0    57.0
SEQN        0
LBXTLG    479
dtype: int64


In [11]:
# Start with demographics
master_df = demo_subset.copy()

# Merge dietary fiber
master_df = master_df.merge(diet_subset, on='SEQN', how='inner')

# Merge glucose
master_df = master_df.merge(glucose_subset, on='SEQN', how='inner')

# Merge HDL
master_df = master_df.merge(hdl_subset, on='SEQN', how='inner')

# Merge Total Cholesterol
master_df = master_df.merge(chol_subset, on='SEQN', how='inner')

# Merge Triglycerides
master_df = master_df.merge(trig_subset, on='SEQN', how='inner')

# Inspect
print(master_df.head())
print(master_df.info())
print(master_df.isnull().sum())

       SEQN  RIAGENDR  RIDAGEYR  DR1TFIBE  LBXGLU  LBDHDD  LBXTC  LBXTLG
0  130378.0       1.0      43.0      10.1   113.0    45.0  264.0   153.0
1  130379.0       1.0      66.0      29.6    99.0    60.0  214.0    86.0
2  130380.0       2.0      44.0      21.1   156.0    49.0  187.0   375.0
3  130386.0       1.0      34.0      25.1   100.0    46.0  183.0   142.0
4  130394.0       1.0      51.0      19.7    88.0    48.0  183.0    57.0
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3996 entries, 0 to 3995
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   SEQN      3996 non-null   float64
 1   RIAGENDR  3996 non-null   float64
 2   RIDAGEYR  3996 non-null   float64
 3   DR1TFIBE  3114 non-null   float64
 4   LBXGLU    3672 non-null   float64
 5   LBDHDD    3546 non-null   float64
 6   LBXTC     3546 non-null   float64
 7   LBXTLG    3517 non-null   float64
dtypes: float64(8)
memory usage: 249.9 KB
None
SEQN          0
RIA

In [12]:
master_df.rename(columns={
    'SEQN': 'ID',
    'RIAGENDR': 'Gender',
    'RIDAGEYR': 'Age',
    'DR1TFIBE': 'Fiber',
    'LBXGLU': 'Glucose',
    'LBDHDD': 'HDL',
    'LBXTC': 'TotalChol',
    'LBXTLG': 'Triglycerides'
}, inplace=True)

# Inspect
print(master_df.head())

         ID  Gender   Age  Fiber  Glucose   HDL  TotalChol  Triglycerides
0  130378.0     1.0  43.0   10.1    113.0  45.0      264.0          153.0
1  130379.0     1.0  66.0   29.6     99.0  60.0      214.0           86.0
2  130380.0     2.0  44.0   21.1    156.0  49.0      187.0          375.0
3  130386.0     1.0  34.0   25.1    100.0  46.0      183.0          142.0
4  130394.0     1.0  51.0   19.7     88.0  48.0      183.0           57.0


In [13]:
# Display top 50 rows
master_df.head(50)

,ID,Gender,Age,Fiber,Glucose,HDL,TotalChol,Triglycerides
0,130378.0,1.0,43.0,10.1,113.0,45.0,264.0,153.0
1,130379.0,1.0,66.0,29.6,99.0,60.0,214.0,86.0
2,130380.0,2.0,44.0,21.1,156.0,49.0,187.0,375.0
3,130386.0,1.0,34.0,25.1,100.0,46.0,183.0,142.0
4,130394.0,1.0,51.0,19.7,88.0,48.0,183.0,57.0
5,130395.0,2.0,33.0,NaN,100.0,NaN,NaN,NaN
6,130396.0,2.0,56.0,NaN,104.0,68.0,214.0,108.0
7,130397.0,2.0,67.0,9.4,138.0,NaN,NaN,NaN
8,130398.0,1.0,47.0,24.7,107.0,41.0,189.0,146.0
9,130402.0,2.0,26.0,NaN,73.0,54.0,135.0,84.0


In [14]:
#Descriptive analysis before handlling the missing values.
#Missing values are excluded from this analysis
master_df.describe()

,ID,Gender,Age,Fiber,Glucose,HDL,TotalChol,Triglycerides
count,3996.000000,3996.000000,3996.000000,3114.000000,3672.000000,3546.000000,3546.000000,3517.000000
mean,136384.084084,1.549550,48.638639,16.254399,107.884532,54.509306,183.126904,115.967302
std,3452.307779,0.497601,20.891412,10.431201,32.476268,14.322738,43.005589,84.368104
min,130378.000000,1.000000,12.000000,0.000000,59.000000,23.000000,62.000000,19.000000
25%,133373.750000,1.000000,31.000000,9.200000,93.000000,44.000000,152.000000,70.000000
50%,136412.500000,2.000000,52.000000,14.200000,100.000000,53.000000,180.000000,96.000000
75%,139339.750000,2.000000,66.000000,20.700000,109.000000,62.000000,210.000000,136.000000
max,142309.000000,2.000000,80.000000,117.300000,561.000000,159.000000,405.000000,1745.000000


In [15]:
# ── CHECK EXTREME VALUES IN EACH VARIABLE ────────────────────────────────────
# We are checking for values that are clinically implausible
# These could be data entry errors or extreme outliers
# that would distort our analysis

# Fiber > 60g
# Recommended daily intake is 25-38g
# Values above 60g from a single 24-hour recall are implausible
print("Fiber > 60g:          ", (master_df['Fiber'] > 60).sum())

# Glucose > 400 mg/dL
# Values above 400 indicate severe uncontrolled diabetes
# Only 3 participants had this — too extreme to keep in regression
print("Glucose > 400:        ", (master_df['Glucose'] > 400).sum())

# Triglycerides > 1000 mg/dL
# Normal range is below 150 mg/dL
# Above 1000 indicates severe hypertriglyceridemia — a rare medical condition
# These 3 values would dominate and skew the entire analysis
print("Triglycerides > 1000: ", (master_df['Triglycerides'] > 1000).sum())

# Total Cholesterol < 100 mg/dL
# Values this low are extremely rare in ambulatory healthy adults
# Could indicate severe malnutrition, liver disease, or data error
print("TotalChol < 100:      ", (master_df['TotalChol'] < 100).sum())

# HDL > 120 mg/dL
# Normal HDL range is 20-100 mg/dL in most adults
# Values above 120 are physiologically unusual and warrant inspection
print("HDL > 120:            ", (master_df['HDL'] > 120).sum())

# HDL < 20 mg/dL
# Values below 20 are extremely rare and likely data errors
# HDL this low is associated with severe metabolic disorders
print("HDL < 20:             ", (master_df['HDL'] < 20).sum())

# Age under 20
# NHANES metabolic research is standardly restricted to adults 20 and older
# Adolescent glucose and lipid reference ranges are physiologically different
# Mixing adolescents with adults would make findings harder to interpret
print("Age under 20:         ", (master_df['Age'] < 20).sum())

Fiber > 60g:           21
Glucose > 400:         3
Triglycerides > 1000:  3
TotalChol < 100:       34
HDL > 120:             4
HDL < 20:              0
Age under 20:          571


In [16]:
# Inspect the low cholesterol participants before deciding
low_chol = master_df[master_df['TotalChol'] < 100]
print(low_chol[['ID', 'Age', 'Gender', 'TotalChol', 'Glucose', 'Triglycerides']].to_string())

            ID   Age  Gender  TotalChol  Glucose  Triglycerides
29    130465.0  77.0     1.0       84.0    126.0           58.0
142   130801.0  56.0     1.0       88.0    103.0           78.0
179   130898.0  13.0     1.0       93.0     80.0           31.0
410   131633.0  13.0     1.0       94.0    102.0           28.0
547   132073.0  79.0     1.0       95.0    108.0           44.0
590   132195.0  14.0     1.0       88.0     90.0           35.0
902   133101.0  45.0     1.0       89.0    122.0          106.0
1030  133484.0  74.0     1.0       86.0    209.0           99.0
1050  133536.0  68.0     1.0       84.0    131.0          114.0
1147  133827.0  67.0     1.0       88.0    110.0           60.0
1295  134280.0  25.0     2.0       87.0     83.0           29.0
1313  134346.0  13.0     1.0       95.0     91.0           37.0
1672  135458.0  30.0     1.0       91.0     96.0            NaN
1686  135499.0  70.0     1.0       94.0    119.0          127.0
1952  136285.0  72.0     1.0       94.0 

In [17]:
import pandas as pd
import numpy as np

# ── STEP 1: CHECK YOUR STARTING POINT ────────────────────────────────────────
# Before removing anything, record how many participants we started with
original_size = len(master_df)
print(f"Original sample size: {original_size}")


# ── STEP 2: REMOVE PARTICIPANTS UNDER 20 ─────────────────────────────────────
# NHANES metabolic research is standardly restricted to adults 20 and older
# Adolescent glucose and lipid reference ranges are physiologically different
# Mixing them with adults would make findings harder to interpret


# ── STEP 3: REMOVE IMPLAUSIBLE FIBER VALUES (> 60g) ──────────────────────────
# Recommended daily fiber intake is 25-38g
# Values above 60g from a single 24-hour recall are likely reporting errors
# We use isnull() condition to KEEP rows where fiber is missing
# We are not removing people for missing data — only for implausible values


# ── STEP 4: REMOVE EXTREME GLUCOSE VALUES (> 400 mg/dL) ──────────────────────
# Only 3 participants had glucose above 400
# These are real but extreme values that would disproportionately
# distort our regression analysis of fiber vs glucose


# ── STEP 5: REMOVE EXTREME TRIGLYCERIDE VALUES (> 1000 mg/dL) ────────────────
# Only 3 participants had triglycerides above 1000
# Severe hypertriglyceridemia is real but these extreme values
# would dominate and skew the entire analysis


# ── STEP 6: REMOVE IMPLAUSIBLE TOTAL CHOLESTEROL VALUES (< 70 mg/dL) ─────────
# After inspecting the 34 low cholesterol participants:
# - Values of 62 and 63 mg/dL are almost certainly data errors
# - Values between 70-99 mg/dL are unusual but clinically defensible
# - We set the cutoff at 70 mg/dL to remove only clear errors


# ── APPLY ALL FILTERS TOGETHER ────────────────────────────────────────────────
master_df_clean = master_df[
    (master_df['Age'] >= 20) &                                              # Adults only
    ((master_df['Fiber'] <= 60) | (master_df['Fiber'].isnull())) &         # Remove implausible fiber
    ((master_df['Glucose'] <= 400) | (master_df['Glucose'].isnull())) &    # Remove extreme glucose
    ((master_df['Triglycerides'] <= 1000) | (master_df['Triglycerides'].isnull())) &  # Remove extreme TG
    ((master_df['TotalChol'] >= 70) | (master_df['TotalChol'].isnull()))   
    # Remove implausible cholesterol. Only 2 value  (62 and 63 are clear data errors 
    # and some other values below 100 are removed by age filter
].copy()


# ── STEP 7: RESET INDEX ───────────────────────────────────────────────────────
# After filtering, row numbers will have gaps (e.g. 0, 2, 5...)
# Resetting gives us clean sequential index (0, 1, 2, 3...)
master_df_clean = master_df_clean.reset_index(drop=True)


# ── STEP 8: VERIFY YOUR RESULTS ───────────────────────────────────────────────
print(f"Original sample:        {original_size}")
print(f"Clean sample:           {len(master_df_clean)}")
print(f"Participants removed:   {original_size - len(master_df_clean)}")

print(f"\nMissing values in clean dataset:")
print(master_df_clean.isnull().sum())

Original sample size: 3996
Original sample:        3996
Clean sample:           3399
Participants removed:   597

Missing values in clean dataset:
ID                 0
Gender             0
Age                0
Fiber            698
Glucose          215
HDL              324
TotalChol        324
Triglycerides    346
dtype: int64


In [18]:
# How many adults have TotalChol between 70-99 in our clean dataset?
low_chol_adults = master_df_clean[
    (master_df_clean['TotalChol'] >= 70) & 
    (master_df_clean['TotalChol'] < 100)
]

print(f"Adults with TotalChol 70-99:   {len(low_chol_adults)}")
print(f"Total clean dataset:           {len(master_df_clean)}")
print(f"Percentage:                    {len(low_chol_adults)/len(master_df_clean)*100:.2f}%")

Adults with TotalChol 70-99:   25
Total clean dataset:           3399
Percentage:                    0.74%


#25 adults (0.74% of sample) had TotalChol between 70-99 mg/dL. Although unusual, this proportion is too small to meaningfully influence regression results. These participants were retained and only the two clearly implausible values of 62 and 63 mg/dL were removed."

In [19]:
# ── CREATE MISSING INDICATOR FLAGS ───────────────────────────────────────────
# For each variable with missing data, we create a new column
# 1 = data is MISSING for that participant
# 0 = data is PRESENT for that participant
# This lets us statistically compare the two groups

master_df_clean['Fiber_missing']   = master_df_clean['Fiber'].isnull().astype(int)
master_df_clean['Glucose_missing'] = master_df_clean['Glucose'].isnull().astype(int)
master_df_clean['HDL_missing']     = master_df_clean['HDL'].isnull().astype(int)


# ── CHECK IF MISSINGNESS IS RELATED TO AGE ───────────────────────────────────
# If people of certain ages are more likely to have missing data,
# missingness is NOT random — it is related to an observed variable (MAR)

print("=== Average Age: Fiber Missing vs Present ===")
print(master_df_clean.groupby('Fiber_missing')['Age'].mean())

print("\n=== Average Age: Glucose Missing vs Present ===")
print(master_df_clean.groupby('Glucose_missing')['Age'].mean())

print("\n=== Average Age: HDL Missing vs Present ===")
print(master_df_clean.groupby('HDL_missing')['Age'].mean())


# ── CHECK IF MISSINGNESS IS RELATED TO GENDER ────────────────────────────────
# Gender = 1 is Male, Gender = 2 is Female in NHANES
# If one gender has more missing data, missingness is related to gender (MAR)

print("\n=== Gender Distribution: Fiber Missing vs Present ===")
print(master_df_clean.groupby('Fiber_missing')['Gender'].value_counts())

print("\n=== Gender Distribution: Glucose Missing vs Present ===")
print(master_df_clean.groupby('Glucose_missing')['Gender'].value_counts())


# ── CHECK HOW MANY VARIABLES EACH PARTICIPANT IS MISSING ─────────────────────
# If the same people are missing fiber AND glucose AND lipids,
# that is a pattern — those participants likely skipped entire
# portions of the NHANES protocol, which is meaningful and not random

master_df_clean['total_missing'] = master_df_clean[
    ['Fiber', 'Glucose', 'HDL', 'TotalChol', 'Triglycerides']
].isnull().sum(axis=1)

print("\n=== How Many Variables Missing Per Participant ===")
print(master_df_clean['total_missing'].value_counts().sort_index())

=== Average Age: Fiber Missing vs Present ===
Fiber_missing
0    55.482044
1    49.293696
Name: Age, dtype: float64

=== Average Age: Glucose Missing vs Present ===
Glucose_missing
0    54.385364
1    51.632558
Name: Age, dtype: float64

=== Average Age: HDL Missing vs Present ===
HDL_missing
0    54.335285
1    53.033951
Name: Age, dtype: float64

=== Gender Distribution: Fiber Missing vs Present ===
Fiber_missing  Gender
0              2.0       1512
               1.0       1189
1              2.0        370
               1.0        328
Name: count, dtype: int64

=== Gender Distribution: Glucose Missing vs Present ===
Glucose_missing  Gender
0                2.0       1765
                 1.0       1419
1                2.0        117
                 1.0         98
Name: count, dtype: int64

=== How Many Variables Missing Per Participant ===
total_missing
0    2448
1     607
2      14
3     105
4     168
5      57
Name: count, dtype: int64


In [20]:
pip install scipy

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [21]:
from scipy import stats

# Fiber
fiber_present = master_df_clean[master_df_clean['Fiber_missing'] == 0]['Age']
fiber_missing = master_df_clean[master_df_clean['Fiber_missing'] == 1]['Age']
t_stat, p_value = stats.ttest_ind(fiber_present, fiber_missing)

print(f"Fiber p-value:   {p_value:.4f}")

# Glucose
glucose_present = master_df_clean[master_df_clean['Glucose_missing'] == 0]['Age']
glucose_missing = master_df_clean[master_df_clean['Glucose_missing'] == 1]['Age']
t_stat2, p_value2 = stats.ttest_ind(glucose_present, glucose_missing)

print(f"Glucose p-value: {p_value2:.4f}")

# HDL
hdl_present = master_df_clean[master_df_clean['HDL_missing'] == 0]['Age']
hdl_missing  = master_df_clean[master_df_clean['HDL_missing'] == 1]['Age']
t_stat3, p_value3 = stats.ttest_ind(hdl_present, hdl_missing)

print(f"HDL p-value:     {p_value3:.4f}")

Fiber p-value:   0.0000
Glucose p-value: 0.0226
HDL p-value:     0.1937


In [22]:
from scipy import stats

# ── CREATE MISSING FLAGS FOR ALL VARIABLES ────────────────────────────────────
# We already created these three earlier
# Now adding the two we missed

master_df_clean['TotalChol_missing']     = master_df_clean['TotalChol'].isnull().astype(int)
master_df_clean['Triglycerides_missing'] = master_df_clean['Triglycerides'].isnull().astype(int)

# ── T-TEST FOR ALL FIVE VARIABLES ─────────────────────────────────────────────
# Testing whether age is significantly different between
# missing and present groups for every variable with missingness

variables = {
    'Fiber':         'Fiber_missing',
    'Glucose':       'Glucose_missing',
    'HDL':           'HDL_missing',
    'TotalChol':     'TotalChol_missing',
    'Triglycerides': 'Triglycerides_missing'
}

print("=== T-test Results: Age vs Missingness ===\n")
for var_name, flag in variables.items():
    present = master_df_clean[master_df_clean[flag] == 0]['Age']
    missing = master_df_clean[master_df_clean[flag] == 1]['Age']
    t_stat, p_value = stats.ttest_ind(present, missing)
    
    significance = "Significant — MAR" if p_value < 0.05 else "Not significant — possibly MCAR"
    
    print(f"{var_name}:")
    print(f"  Mean age present: {present.mean():.1f}")
    print(f"  Mean age missing: {missing.mean():.1f}")
    print(f"  P-value:          {p_value:.4f}  →  {significance}")
    print()

=== T-test Results: Age vs Missingness ===

Fiber:
  Mean age present: 55.5
  Mean age missing: 49.3
  P-value:          0.0000  →  Significant — MAR

Glucose:
  Mean age present: 54.4
  Mean age missing: 51.6
  P-value:          0.0226  →  Significant — MAR

HDL:
  Mean age present: 54.3
  Mean age missing: 53.0
  P-value:          0.1937  →  Not significant — possibly MCAR

TotalChol:
  Mean age present: 54.3
  Mean age missing: 53.0
  P-value:          0.1937  →  Not significant — possibly MCAR

Triglycerides:
  Mean age present: 54.4
  Mean age missing: 52.9
  P-value:          0.1381  →  Not significant — possibly MCAR



## Missingness Diagnosis — Complete

Independent samples t-tests were conducted to assess whether missingness was related to age across all five variables with missing data.

Fiber (p < 0.001) and Glucose (p = 0.023) showed significant age-related missingness — younger participants were more likely to have missing dietary recall and fasting glucose data. These variables are classified as MAR.

HDL, TotalChol, and Triglycerides showed no significant age relationship (p > 0.05). Notably, HDL and TotalChol had identical missingness counts (324), confirming the same participants missed both — consistent with blood draw 
non-completion rather than age-related dropout. These variables are classified as possibly MCAR.

Given the MAR classification for primary variables, KNN imputation will be used as it leverages observed variables to intelligently estimate missing values.

In [23]:
# ── CONFIRM MISSING VALUES BEFORE IMPUTATION ─────────────────────────────────
# This is our baseline — we will compare against this after imputation
# to confirm everything was handled correctly

print("Missing values BEFORE imputation:")
print(master_df_clean.isnull().sum())
print(f"\nTotal missing values: {master_df_clean.isnull().sum().sum()}")

Missing values BEFORE imputation:
ID                         0
Gender                     0
Age                        0
Fiber                    698
Glucose                  215
HDL                      324
TotalChol                324
Triglycerides            346
Fiber_missing              0
Glucose_missing            0
HDL_missing                0
total_missing              0
TotalChol_missing          0
Triglycerides_missing      0
dtype: int64

Total missing values: 1907


In [24]:
pip install scikit-learn

Defaulting to user installation because normal site-packages is not writeable
Note: you may need to restart the kernel to use updated packages.


In [25]:
# sklearn contains both the scaler and KNN imputer we need
import sklearn
print(sklearn.__version__)

1.6.1


In [26]:
from sklearn.preprocessing import StandardScaler
from sklearn.impute import KNNImputer

# ── STEP 1: SELECT ONLY THE COLUMNS WE WILL IMPUTE ───────────────────────────
# We use the five clinical variables plus age and gender
# Age and gender are included because they help KNN find similar participants
# They have no missing values so they purely guide the similarity calculation

cols_to_impute = ['Age', 'Gender', 'Fiber', 'Glucose', 
                  'HDL', 'TotalChol', 'Triglycerides']

df_to_impute = master_df_clean[cols_to_impute].copy()

print("Shape of data going into imputation:", df_to_impute.shape)
print("\nMissing values before standardization:")
print(df_to_impute.isnull().sum())

Shape of data going into imputation: (3399, 7)

Missing values before standardization:
Age                0
Gender             0
Fiber            698
Glucose          215
HDL              324
TotalChol        324
Triglycerides    346
dtype: int64


In [27]:
from sklearn.preprocessing import StandardScaler
from sklearn.impute import KNNImputer

# ── STEP 1: SELECT ONLY THE COLUMNS WE WILL IMPUTE ───────────────────────────
# We use the five clinical variables plus age and gender
# Age and gender are included because they help KNN find similar participants
# They have no missing values so they purely guide the similarity calculation

cols_to_impute = ['Age', 'Gender', 'Fiber', 'Glucose', 
                  'HDL', 'TotalChol', 'Triglycerides']

df_to_impute = master_df_clean[cols_to_impute].copy()

# ── STEP 2: STANDARDIZE THE DATA ─────────────────────────────────────────────
# Converts every variable to mean=0 and std=1
# So no single variable dominates the distance calculation

scaler = StandardScaler()
df_scaled = scaler.fit_transform(df_to_impute)

# ── STEP 3: APPLY KNN IMPUTATION ─────────────────────────────────────────────
# Finds 5 most similar participants for each missing value
# and averages their values to fill the gap

imputer = KNNImputer(n_neighbors=5, weights='uniform')
df_imputed_scaled = imputer.fit_transform(df_scaled)

# ── STEP 4: REVERSE THE STANDARDIZATION ──────────────────────────────────────
# Converts everything back to original units
# Glucose back to mg/dL, Fiber back to grams etc.

df_imputed_original = scaler.inverse_transform(df_imputed_scaled)

# ── STEP 5: PUT IMPUTED VALUES BACK INTO A DATAFRAME ─────────────────────────
# Converts numpy array back to pandas dataframe with correct column names

df_imputed = pd.DataFrame(df_imputed_original, columns=cols_to_impute)

# ── STEP 6: VERIFY IMPUTATION WORKED ─────────────────────────────────────────
print("Missing values after imputation:")
print(df_imputed.isnull().sum())
print(f"\nTotal missing values: {df_imputed.isnull().sum().sum()}")

Missing values after imputation:
Age              0
Gender           0
Fiber            0
Glucose          0
HDL              0
TotalChol        0
Triglycerides    0
dtype: int64

Total missing values: 0


/Users/ushagartoula/Library/Python/3.9/lib/python/site-packages/sklearn/utils/extmath.py:203: RuntimeWarning: divide by zero encountered in matmul
  ret = a @ b
/Users/ushagartoula/Library/Python/3.9/lib/python/site-packages/sklearn/utils/extmath.py:203: RuntimeWarning: overflow encountered in matmul
  ret = a @ b
/Users/ushagartoula/Library/Python/3.9/lib/python/site-packages/sklearn/utils/extmath.py:203: RuntimeWarning: invalid value encountered in matmul
  ret = a @ b


In [28]:
# ── CHECK STANDARD DEVIATION OF EACH VARIABLE ────────────────────────────────
# A standard deviation of zero or near zero means all values are identical
# This would cause divide by zero during standardization

print("Standard deviation of each variable:")
print(df_to_impute.std())

print("\nUnique values in Gender:")
print(df_to_impute['Gender'].value_counts())

Standard deviation of each variable:
Age              17.139530
Gender            0.497182
Fiber             9.397004
Glucose          31.790781
HDL              14.730301
TotalChol        42.826678
Triglycerides    75.604546
dtype: float64

Unique values in Gender:
Gender
2.0    1882
1.0    1517
Name: count, dtype: int64


In [29]:
# ── COMPARE DESCRIPTIVE STATS BEFORE AND AFTER IMPUTATION ────────────────────
# We check that imputed values fall within clinically reasonable ranges
# If imputation worked correctly, means and distributions should
# be very similar before and after

print("=== BEFORE IMPUTATION ===")
print(master_df_clean[['Fiber', 'Glucose', 'HDL', 
                         'TotalChol', 'Triglycerides']].describe().round(2))

print("\n=== AFTER IMPUTATION ===")
print(df_imputed[['Fiber', 'Glucose', 'HDL', 
                   'TotalChol', 'Triglycerides']].describe().round(2))

=== BEFORE IMPUTATION ===
         Fiber  Glucose      HDL  TotalChol  Triglycerides
count  2701.00  3184.00  3075.00    3075.00        3053.00
mean     16.08   109.30    54.95     187.34         119.47
std       9.40    31.79    14.73      42.83          75.60
min       0.00    59.00    23.00      80.00          25.00
25%       9.40    94.00    44.00     157.00          73.00
50%      14.20   101.00    53.00     185.00          99.00
75%      20.80   111.00    63.00     214.00         142.00
max      57.20   385.00   159.00     405.00         800.00

=== AFTER IMPUTATION ===
         Fiber  Glucose      HDL  TotalChol  Triglycerides
count  3399.00  3399.00  3399.00    3399.00        3399.00
mean     15.86   109.34    54.86     187.11         119.20
std       8.57    31.05    14.23      41.42          72.54
min       0.00    59.00    23.00      80.00          25.00
25%      10.20    94.00    45.00     159.00          75.00
50%      14.30   101.60    53.00     185.00         101.00
75% 

In [30]:
# ── PUT IMPUTED VALUES BACK INTO MASTER DATASET ──────────────────────────────
# We replace the original columns that had missing values
# with the imputed versions from df_imputed
# ID column is carried over from master_df_clean unchanged

master_df_final = master_df_clean[['ID']].copy()

# Add all imputed columns
for col in cols_to_impute:
    master_df_final[col] = df_imputed[col].values

# ── FINAL VERIFICATION ────────────────────────────────────────────────────────
print("Final dataset shape:", master_df_final.shape)
print("\nMissing values in final dataset:")
print(master_df_final.isnull().sum())
print(f"\nTotal missing values: {master_df_final.isnull().sum().sum()}")

Final dataset shape: (3399, 8)

Missing values in final dataset:
ID               0
Age              0
Gender           0
Fiber            0
Glucose          0
HDL              0
TotalChol        0
Triglycerides    0
dtype: int64

Total missing values: 0


## Missing Data Handling — Summary

Dataset was cleaned and missing values handled in the following steps:

1. OUTLIER REMOVAL
   597 participants removed including 571 under age 20 and 26 with
   clinically implausible values across fiber, glucose, triglycerides,
   and total cholesterol.

2. MISSINGNESS DIAGNOSIS
   Independent samples t-tests revealed fiber (p<0.001) and glucose 
   (p=0.023) missingness were significantly associated with age — 
   classified as MAR. HDL, TotalChol and Triglycerides showed no 
   significant age relationship (p>0.05) — classified as MCAR.

3. IMPUTATION METHOD
   KNN imputation (k=5) was applied using age, gender, and all 
   clinical variables as features. Variables were standardized 
   before imputation and back-transformed afterward.

4. VALIDATION
   Post-imputation means were virtually identical to pre-imputation 
   means. No impossible values were created. All 3399 participants 
   now have complete data across all variables.

In [31]:
# ── SAVE YOUR FINAL CLEAN DATASET ───────────────────────────────────────────

master_df_final.to_csv('nhanes_clean_imputed.csv', index=False)
print("Dataset saved successfully")

Dataset saved successfully
